lib/crypto-algorithms/megolm.js

```diff
@@ -0,0 +1,171 @@
+/*
+Copyright 2015, 2016 OpenMarket Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+"use strict";
+
+/**
+ * Defines m.olm encryption/decryption
+ *
+ * @module crypto-algorithms/olm
+ */
+
+var q = require("q");
+
+var utils = require("../utils");
+var base = require("./base");
+
+var MEGOLM_ALGORITHM = "m.megolm.v1.aes-sha2";
+
+/**
+ * Megolm encryption implementation
+ *
+ * @constructor
+ * @extends {module:crypto-algorithms/base.EncryptionAlgorithm}
+ *
+ * @param {string} deviceId The identifier for this device.
+ * @param {module:crypto} crypto crypto core
+ * @param {module:OlmDevice} olmDevice olm.js wrapper
+ *
+ */
+function MegolmEncryption(deviceId, crypto, olmDevice) {
+    base.EncryptionAlgorithm.call(this, deviceId, crypto, olmDevice);
+    this._outboundSessionId = null;
+}
+utils.inherits(MegolmEncryption, base.EncryptionAlgorithm);
+
+/**
+ * @inheritdoc
+ * @param {string[]} roomMembers list of currently-joined users in the room
+ * @return {module:client.Promise} Promise which resolves when setup is complete
+ */
+MegolmEncryption.prototype.initRoomEncryption = function(roomMembers) {
+    // nothing to do here.
+    return q();
+};
+
+
+/**
+ * @private
+ * @param {string} roomId
+ */
+MegolmEncryption.prototype._ensureOutboundSession = function(roomId) {
+    if (this._outboundSessionId) {
+        return;
+    }
+    var session_id = this._olmDevice.createOutboundGroupSession();
+    this._outboundSessionId = session_id;
+
+    var key = this._olmDevice.getOutboundGroupSessionKey(session_id);
+
+    // TODO: initiate key-sharing
+
+    console.log(
+        'Created outbound session. Add with window.mxMatrixClientPeg.' +
+            'matrixClient._crypto._olmDevice.addInboundGroupSession("' +
+            [
+                roomId, this._olmDevice.deviceCurve25519Key, session_id,
+                key.key, key.chain_index
+            ].join('", "') +
+            '")'
```

Comment:
yeah, but that doesn't particularly help with the "-wrapping of the arguments to addInboundGroupSession.